keylime/src/device_id.rs

```diff
@@ -0,0 +1,535 @@
+            .map_err(DeviceIDBuilderError::Template)
+    }
+
+    /// Get the IAK from the given handle and password or recreate following the given template
+    ///
+    /// If a handle has been set, try to obtain the IAK from the handle
+    /// If there is a configured IAK password, add the password to the handle
+    ///
+    /// If the handle is empty, recreate the IAK using the provided algorithms
+    fn get_iak(
+        &mut self,
+        tpm_ctx: &mut tpm::Context,
+    ) -> Result<IAKResult, DeviceIDBuilderError> {
+        let (asym_alg, hash_alg) = self.get_iak_template()?;
+        match self.iak_handle {
+            Some(handle) => {
+                if handle.trim().is_empty() {
+                    info!("Recreating IAK.");
+                    tpm_ctx
+                        .create_iak(asym_alg, hash_alg)
+                        .map_err(DeviceIDBuilderError::IAKCreate)
+                } else {
+                    let password = self.iak_password.unwrap_or("").trim();
+                    info!("Collecting persisted IAK.");
+                    tpm_ctx
+                        .iak_from_handle(handle.trim(), password)
+                        .map_err(DeviceIDBuilderError::IAKFromHandle)
+                }
+            }
+            None => Err(DeviceIDBuilderError::IAKHandleNotSet),
+        }
+    }
+
+    /// Get the IDevID from the given handle and password or recreate following the given template
+    ///
+    /// If a handle has been set, try to obtain the IDevID from the handle
+    /// If there is a configured IDevID password, add the password to the handle
+    ///
+    /// If the handle is empty, recreate the IDevId using the provided algorithms
+    fn get_idevid(
+        &mut self,
+        tpm_ctx: &mut tpm::Context,
+    ) -> Result<IDevIDResult, DeviceIDBuilderError> {
+        let (asym_alg, hash_alg) = self.get_idevid_template()?;
+        match self.idevid_handle {
+            Some(handle) => {
+                if handle.trim().is_empty() {
+                    info!("Recreating IDevID.");
+                    tpm_ctx
+                        .create_idevid(asym_alg, hash_alg)
+                        .map_err(DeviceIDBuilderError::IDevIDCreate)
+                } else {
+                    let password = self.idevid_password.unwrap_or("").trim();
+                    info!("Collecting persisted IDevID.");
+                    tpm_ctx
+                        .idevid_from_handle(handle.trim(), password.trim())
+                        .map_err(DeviceIDBuilderError::IDevIDFromHandle)
+                }
+            }
+            None => Err(DeviceIDBuilderError::IDevIDHandleNotSet),
+        }
+    }
+
+    /// Get the IAK certificate
+    /// If the iak_cert is not set, try to load the certificate from the iak_cert_path and cache
+    /// the loaded certificate
+    fn get_iak_cert(&mut self) -> Result<&X509, DeviceIDBuilderError> {
```

Comment:
I think this could be an option, with it only failing if a path is set but cert can't be found, and on empty string it is None. See pubkey cert key check comment for more detail.